specs/interop.md

```diff
@@ -0,0 +1,708 @@
+
+An explicit `_destination` chain and `nonce` are used to ensure that the message can only be played on a single remote
+chain a single time. The `_destination` is enforced to not be the local chain to avoid edge cases.
+
+```solidity
+function sendMessage(uint256 _destination, address _target, bytes calldata _message) external payable {
+    require(_destination != block.chainid);
+
+    bytes memory data = abi.encodeCall(L2ToL2CrossDomainMessenger.relayMessage, (_destination, messageNonce(), msg.sender, _target, msg.value, _message));
+    emit SentMessage(data);
+    nonce++;
+}
+```
+
+##### Relaying Messages
+
+When relaying a message through the `L2ToL2CrossDomainMessenger`, it is important to require that
+the `_destination` equal to `block.chainid` to ensure that the message is only valid on a single
+chain. The hash of the message is used for replay protection.
+
+It is important to ensure that the source chain is in the dependency set of the destination chain, otherwise
+it is possible to send a message that is not playable.
+
+```solidity
+function relayMessage(uint256 _destination, uint256 _nonce, address _sender, address _target, uint256 _value, bytes memory _message) external {
+    require(msg.sender == address(CROSS_L2_INBOX));
+    require(_destination == block.chainid);
+    require(CROSS_L2_INBOX.origin() == address(this));
+    require(_target != address(this));
+
+    bytes32 messageHash = keccak256(abi.encode(_destination, _nonce, _sender, _target, _value, _message));
+    require(sentMessages[messageHash] == false);
+
+    assembly {
+      tstore(CROSS_DOMAIN_MESSAGE_SENDER_SLOT, _sender)
+    }
+
+    bool success = SafeCall.call({
+       _target: _target,
+       _value: _value,
+       _calldata: _message
+    });
+
+    require(success);
+
+    sentMessages[messageHash] = true;
+}
+```
+
+### L1Block
+
+| Constant         | Value                   |
+|----------------- | ----------------------- |
+| Address | `0x4200000000000000000000000000000000000015` |
+
+The `L1Block` contract is updated to include the set of allowed chains. The L1 Atrributes transaction
+sets the set of allowed chains. The `L1Block` contract MUST provide a public getter to check is a particular
+chain is in the dependency set called `isInDependencySet(uint256)`. This function MUST return true when
+the chain's chain id is passed in as an argument.
+
+The `setL1BlockValuesInterop()` function MUST be called on every block after the interop upgrade block.
+The interop upgrade block itself MUST include a call to `setL1BlockValuesEcotone`.
+
+## SystemConfig
+
+The `SystemConfig` is updated to manage the dependency set. The chain operator can add or remove
+chains from the dependency set through the `SystemConfig`. A new `ConfigUpdate` event `UpdateType`
+enum is added that corresponds to a change in the dependency set.
+
+The `SystemConfig` MUST enforce that the maximum size of the dependency set is `type(uint8).max` or 255.
```

Comment:
Where does this number come from, and will be upgradeable in the future?

Reply:
I'd be curious to know if there has been consideration as to how this evolves over time, also since the size of the data has to be small enough. 

Related questions:
- The values will change over time, right, so if we're at 255 and we want to add a new one, it will be possible to remove one and add a new one? If so, this is kind of like the "current" dependency set.
- In a future where we want, say, 1000 dependant chains for whatever reason, is the approach going to involve merklizing the list? Does that break anything?

Reply:
This number is chosen because it limits the size of the L1 attributes tx. The security considerations section talks about this a bit, but the tldr is that we cannot consume more than the block gas limit with force include txs and the cost of promoting txs from unsafe to safe.

This number feels quite sane for the first version, especially since the process of verifying the initiating messages in the derivation pipeline will be expensive. It will use a lot of bandwidth. We will figure out how to make it cheaper over time but the first implementation will not be super scalable. Anything over 10 seems like it would be kind of crazy but will need to benchmark to know for sure.

If you want many dependent chains, you can have a multi hub and spoke kind of model. All transitive deps technically count, so just because you cannot directly accept incoming messages from a chain doesn't mean that you cannot, you will just need a hop in between. In the future we can consider relaxing this restriction depending on how cheap we can make the verification in the derivation pipeline. We have some fun ideas to use kzg since the witness size is constant, perhaps worldcoin would like to help with the cryptography?